File: backend/app/api/crypto_fees.py
from datetime import date
from typing import Annotated

from fastapi import APIRouter, Depends, Query, Response, status
from sqlmodel import Session

from app.db.session import get_session
from app.models.crypto_fee import CryptoFeeType
from app.schemas.crypto_fee import (
    CryptoFeeCreate,
    CryptoFeeRead,
    CryptoFeeUpdate,
)
from app.services.crypto_fee_service import (
    create_crypto_fee,
    delete_crypto_fee,
    get_crypto_fee_read,
    list_crypto_fees,
    update_crypto_fee,
)

router = APIRouter(prefix="/crypto-fees", tags=["crypto-fees"])


@router.get("", response_model=list[CryptoFeeRead])
def get_crypto_fees(
    session: Annotated[Session, Depends(get_session)],
    portfolio_id: int | None = None,
    asset_id: int | None = None,
    fee_type: CryptoFeeType | None = None,
    from_date: date | None = None,
    to_date: date | None = None,
) -> list[CryptoFeeRead]:
    return list_crypto_fees(
        session,
        portfolio_id=portfolio_id,
        asset_id=asset_id,
        fee_type=fee_type,
        from_date=from_date,
        to_date=to_date,
    )


@router.post("", response_model=CryptoFeeRead, status_code=status.HTTP_201_CREATED)
def post_crypto_fee(
    payload: CryptoFeeCreate,
    session: Annotated[Session, Depends(get_session)],
) -> CryptoFeeRead:
    return create_crypto_fee(session, payload)


@router.get("/{fee_id}", response_model=CryptoFeeRead)
def get_crypto_fee(
    fee_id: int,
    session: Annotated[Session, Depends(get_session)],
) -> CryptoFeeRead:
    return get_crypto_fee_read(session, fee_id)


@router.patch("/{fee_id}", response_model=CryptoFeeRead)
def patch_crypto_fee(
    fee_id: int,
    payload: CryptoFeeUpdate,
    session: Annotated[Session, Depends(get_session)],
) -> CryptoFeeRead:
    return update_crypto_fee(session, fee_id, payload)


@router.delete("/{fee_id}", status_code=status.HTTP_204_NO_CONTENT)
def remove_crypto_fee(
    fee_id: int,
    session: Annotated[Session, Depends(get_session)],
) -> Response:
    delete_crypto_fee(session, fee_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: frontend/src/lib/api/cryptoFees.ts
import { API_BASE_URL } from './config';
import { apiFetch } from './http';

type ApiFetcher = typeof apiFetch;

export type CryptoFeeType = 'purchase' | 'transfer';

export type CryptoFee = {
  id: number;
  portfolio_id: number;
  asset_id: number;
  fee_type: CryptoFeeType;
  fee_date: string;
  quantity_moved: number;
  fee_quantity_btc: number;
  quote_brl: number;
  fx_rate: number;
  notes?: string | null;
  symbol: string;
  asset_name: string;
  final_quantity_after_fee: number;
  fee_value_brl: number;
  fee_value_usd: number;
  fee_percent: number;
};

export type CryptoFeeCreate = {
  portfolio_id: number;
  asset_id: number;
  fee_type: CryptoFeeType;
  fee_date: string;
  quantity_moved: number;
  fee_quantity_btc: number;
  quote_brl: number;
  fx_rate: number;
  notes?: string | null;
};

export type CryptoFeeUpdate = Partial<CryptoFeeCreate>;

export type CryptoFeeListParams = {
  portfolio_id?: number;
  asset_id?: number;
  fee_type?: CryptoFeeType;
  from_date?: string;
  to_date?: string;
};

export type BitcoinPositionSummary = {
  asset_id: number | null;
  symbol: string | null;
  name: string | null;
  quantity: number | null;
  average_price_brl: number | null;
  average_price_usd: number | null;
  total_invested_brl: number | null;
  quote_brl: number | null;
  quote_usd: number | null;
  current_value_brl: number | null;
  profit_brl: number | null;
  profit_percent: number | null;
};

export type BitcoinRebalanceSummary = {
  target_value_brl: number | null;
  missing_value_brl: number | null;
  above_target_brl: number | null;
  rebalance_action: string | null;
};

export type BitcoinSnapshot = {
  portfolio_id: number;
  position: BitcoinPositionSummary;
  rebalance: BitcoinRebalanceSummary;
  total_fees_brl: number;
  total_fees_usd: number;
  profit_after_fees_brl: number | null;
  appreciation_after_fees_percent: number | null;
  transfer_ledger_final_btc: number;
  transfer_ledger_count: number;
};

async function parseResponse<T>(response: Response): Promise<T> {
  if (!response.ok) {
    throw new Error(await response.text());
  }
  return response.json() as Promise<T>;
}

function buildQuery(params?: CryptoFeeListParams): string {
  if (!params) {
    return '';
  }
  const search = new URLSearchParams();
  if (params.portfolio_id != null) {
    search.set('portfolio_id', String(params.portfolio_id));
  }
  if (params.asset_id != null) {
    search.set('asset_id', String(params.asset_id));
  }
  if (params.fee_type) {
    search.set('fee_type', params.fee_type);
  }
  if (params.from_date) {
    search.set('from_date', params.from_date);
  }
  if (params.to_date) {
    search.set('to_date', params.to_date);
  }
  const query = search.toString();
  return query ? `?${query}` : '';
}

export async function listCryptoFees(
  params?: CryptoFeeListParams,
  fetcher: ApiFetcher = apiFetch
): Promise<CryptoFee[]> {
  const response = await fetcher(`${API_BASE_URL}/crypto-fees${buildQuery(params)}`);
  return parseResponse(response);
}

export async function createCryptoFee(
  payload: CryptoFeeCreate,
  fetcher: ApiFetcher = apiFetch
): Promise<CryptoFee> {
  const response = await fetcher(`${API_BASE_URL}/crypto-fees`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload)
  });
  return parseResponse(response);
}

export async function updateCryptoFee(
  id: number,
  payload: CryptoFeeUpdate,
  fetcher: ApiFetcher = apiFetch
): Promise<CryptoFee> {
  const response = await fetcher(`${API_BASE_URL}/crypto-fees/${id}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload)
  });
  return parseResponse(response);
}

export async function deleteCryptoFee(id: number, fetcher: ApiFetcher = apiFetch): Promise<void> {
  const response = await fetcher(`${API_BASE_URL}/crypto-fees/${id}`, { method: 'DELETE' });
  if (!response.ok) {
    throw new Error(await response.text());
  }
}

export async function getBitcoinSnapshot(
  portfolioId: number,
  assetId?: number,
  fetcher: ApiFetcher = apiFetch
): Promise<BitcoinSnapshot> {
  const search = assetId != null ? `?asset_id=${assetId}` : '';
  const response = await fetcher(
    `${API_BASE_URL}/portfolios/${portfolioId}/bitcoin-snapshot${search}`
  );
  return parseResponse(response);
}
